File: android-native/app/src/main/java/com/example/kpkn/domain/nutrition/NutritionHeuristicEstimator.kt
package com.example.kpkn.domain.nutrition

/**
 * NutritionHeuristicEstimator — Keyword-based macro estimator for foods not in the database.
 *
 * Used as fallback when:
 *   - Mode is BASIC (rules-only, no AI)
 *   - PRO mode but Qwen model not loaded / timeout
 *   - AI returned items without nutritionPer100g for some entries
 *
 * Precision: ±20–30%. Tagged as AnalysisSource.LOCAL_HEURISTIC so the UI shows "Heur. ~".
 * Values are per 100g.
 */

data class NutritionProfile(
    val calories: Double,
    val protein: Double,
    val carbs: Double,
    val fats: Double,
)

// ─── Profiles per 100g ───────────────────────────────────────────────────────

private val LEAN_PROTEIN    = NutritionProfile(165.0, 31.0,  0.0,  3.6)
private val FATTY_PROTEIN   = NutritionProfile(220.0, 26.0,  0.0, 12.0)
private val FISH_LEAN       = NutritionProfile(120.0, 22.0,  0.0,  3.0)
private val FISH_FATTY      = NutritionProfile(200.0, 20.0,  0.0, 13.0)
private val EGG_PROFILE     = NutritionProfile(155.0, 13.0,  1.0, 11.0)
private val LEGUME          = NutritionProfile(116.0,  9.0, 20.0,  0.4)
private val GRAIN_COOKED    = NutritionProfile(130.0,  3.0, 27.0,  0.5)
private val GRAIN_DRY       = NutritionProfile(360.0, 13.0, 68.0,  6.0)
private val PASTA_COOKED    = NutritionProfile(131.0,  5.0, 25.0,  1.1)
private val BREAD           = NutritionProfile(265.0,  9.0, 49.0,  3.0)
private val STARCHY_VEG     = NutritionProfile( 85.0,  1.9, 20.0,  0.1)
private val DAIRY_LIQUID    = NutritionProfile( 62.0,  3.2,  4.8,  3.4)
private val DAIRY_THICK     = NutritionProfile( 97.0,  9.0,  3.9,  5.0)  // yogurt / cottage
private val CHEESE          = NutritionProfile(350.0, 22.0,  2.0, 28.0)
private val OIL_FAT         = NutritionProfile(780.0,  0.0,  0.0, 87.0)
private val NUTS_SEEDS      = NutritionProfile(580.0, 20.0, 20.0, 50.0)
private val FRUIT           = NutritionProfile( 60.0,  0.8, 14.0,  0.3)
private val VEGETABLE       = NutritionProfile( 28.0,  2.0,  5.0,  0.3)
private val PROTEIN_POWDER  = NutritionProfile(380.0, 80.0,  8.0,  5.0)
private val SUGAR_SWEET     = NutritionProfile(380.0,  0.5, 90.0,  0.2)
private val PROCESSED_MEAT  = NutritionProfile(240.0, 16.0,  2.0, 19.0)
private val SAUCE_DRESSING  = NutritionProfile(250.0,  1.5, 15.0, 20.0)
private val MIXED_DISH      = NutritionProfile(160.0, 10.0, 16.0,  6.0)  // generic fallback

// ─── Keyword → Profile mapping ───────────────────────────────────────────────
// Rules are checked in order — more specific rules first.

private val KEYWORD_PROFILES: List<Pair<List<String>, NutritionProfile>> = listOf(

    // ── Suplementos (antes que cualquier otro match) ──────────────────────────
    listOf(
        "whey", "proteína en polvo", "proteina en polvo",
        "caseína", "caseina", "suplemento proteico", "mass gainer", "gainers"
    ) to PROTEIN_POWDER,

    // ── Proteínas magras ──────────────────────────────────────────────────────
    listOf(
        "pechuga de pollo", "pechuga de pavo", "pollo a la plancha",
        "pechuga", "pechuga cocida", "pollo cocido", "pavo cocido",
        "clara de huevo", "claras de huevo", "claras",
        "atún", "atun", "merluza", "tilapia", "corvina", "reineta",
        "camarón", "camaron", "langostino", "pulpo", "calamar",
        "pollo", "pavo"
    ) to LEAN_PROTEIN,

    // ── Proteínas con grasa ───────────────────────────────────────────────────
    listOf(
        "carne molida", "carne de vacuno", "carne de res", "carne de cerdo",
        "asado", "costilla", "lomo vetado", "punta de ganso", "tapapecho",
        "paleta", "pierna de cerdo", "chuleta", "filete", "lomo",
        "vacuno", "res", "cerdo", "carne"
    ) to FATTY_PROTEIN,

    // ── Pescado graso ─────────────────────────────────────────────────────────
    listOf("salmón", "salmon", "sardina", "caballa", "atún en aceite", "jurel") to FISH_FATTY,

    // ── Pescado magro ─────────────────────────────────────────────────────────
    listOf("pescado", "congrio", "lenguado", "bacalao") to FISH_LEAN,

    // ── Huevo ─────────────────────────────────────────────────────────────────
    listOf("huevo duro", "huevo cocido", "huevo frito", "huevo revuelto",
        "huevos duros", "huevo", "huevos") to EGG_PROFILE,

    // ── Embutidos / procesados ────────────────────────────────────────────────
    listOf(
        "jamón", "jamon", "salchicha", "vienesa", "longaniza",
        "mortadela", "salame", "cecina", "tocino", "panceta"
    ) to PROCESSED_MEAT,

    // ── Legumbres ─────────────────────────────────────────────────────────────
    listOf(
        "lentejas", "lenteja", "garbanzos", "garbanzo",
        "porotos", "poroto", "frijoles", "frijol",
        "habas", "edamame", "soya"
    ) to LEGUME,

    // ── Pasta (cocida) ────────────────────────────────────────────────────────
    listOf(
        "pasta", "fideos", "spaghetti", "espagueti", "tallarín", "tallarines",
        "penne", "macarrón", "macarrones", "lasaña"
    ) to PASTA_COOKED,

    // ── Avena cruda / granola / cereales secos ────────────────────────────────
    listOf("avena", "granola", "müesli", "muesli", "cereal") to GRAIN_DRY,

    // ── Arroz y granos cocidos ────────────────────────────────────────────────
    listOf(
        "arroz blanco", "arroz integral", "arroz",
        "quinoa", "cebada cocida", "mijo cocido", "trigo bulgur"
    ) to GRAIN_COOKED,

    // ── Pan / masas ───────────────────────────────────────────────────────────
    listOf(
        "marraqueta", "hallulla", "baguette", "pan de molde",
        "pan integral", "pan blanco", "pan", "tortilla", "arepa", "pita"
    ) to BREAD,

    // ── Vegetales con almidón ─────────────────────────────────────────────────
    listOf(
        "papa", "papas", "batata", "camote", "yuca", "ñame", "taro"
    ) to STARCHY_VEG,

    // ── Queso ─────────────────────────────────────────────────────────────────
    listOf(
        "queso cheddar", "queso gauda", "queso mantecoso", "queso fresco",
        "queso crema", "queso", "mozzarella", "parmesano"
    ) to CHEESE,

    // ── Yogurt / cottage (antes que "leche") ──────────────────────────────────
    listOf(
        "yogurt griego", "yogur griego", "yogurt natural", "yogur natural",
        "yogurt", "yogur", "queso cottage"
    ) to DAIRY_THICK,

    // ── Leche ─────────────────────────────────────────────────────────────────
    listOf("leche entera", "leche descremada", "leche semi", "leche") to DAIRY_LIQUID,

    // ── Aceites y grasas puras ────────────────────────────────────────────────
    listOf(
        "aceite de oliva", "aceite de coco", "aceite vegetal", "aceite",
        "mantequilla", "manteca", "ghee", "margarina"
    ) to OIL_FAT,

    // ── Nueces y semillas ─────────────────────────────────────────────────────
    listOf(
        "almendra", "almendras", "nuez", "nueces", "maní", "mani",
        "pecan", "pistache", "semillas de chia", "semillas de linaza",
        "semillas de girasol", "semillas de zapallo", "semilla", "chía", "chia"
    ) to NUTS_SEEDS,

    // ── Frutas ────────────────────────────────────────────────────────────────
    listOf(
        "manzana", "plátano", "platano", "banana", "naranja", "uva",
        "pera", "durazno", "mango", "papaya", "sandía", "sandia",
        "melón", "melon", "piña", "pina", "frutilla", "frambuesa",
        "arándano", "arandano", "kiwi", "ciruela", "damasco",
        "maracuyá", "lúcuma", "fruta"
    ) to FRUIT,

    // ── Verduras ──────────────────────────────────────────────────────────────
    listOf(
        "lechuga", "espinaca", "brócoli", "brocoli", "zanahoria",
        "tomate", "cebolla", "pepino", "zapallo", "pimentón", "pimenton",
        "apio", "rúcula", "repollo", "coliflor", "poroto verde",
        "choclo", "betarraga", "champiñon", "verdura", "vegetal",
        "acelga", "alcachofa", "berenjena", "zucchini", "pepinillo"
    ) to VEGETABLE,

    // ── Azúcar / dulces ───────────────────────────────────────────────────────
    listOf(
        "azúcar", "azucar", "miel", "sirope", "jarabe",
        "mermelada", "dulce de leche", "manjar", "chocolate"
    ) to SUGAR_SWEET,

    // ── Salsas / aderezos ─────────────────────────────────────────────────────
    listOf(
        "mayonesa", "ketchup", "mostaza", "salsa de tomate",
        "salsa de soya", "salsa", "aderezo", "vinagreta"
    ) to SAUCE_DRESSING,
)

// ─── Public API ───────────────────────────────────────────────────────────────

/**
 * Estimates macros per 100g from the food name using keyword matching.
 * Returns null if no keyword matched (unknown food category).
 * Confidence: ~0.45 (rough estimate, ±20–30%).
 *
 * Also detects cooking method keywords and boosts fats/kcal accordingly.
 */
fun estimateNutritionByKeyword(foodName: String): NutritionProfile? {
    val lower = foodName.trim().lowercase()

    // Detect cooking method indicators in the food name
    val cookingBoost = detectCookingBoost(lower)

    for ((keywords, profile) in KEYWORD_PROFILES) {
        if (keywords.any { lower.contains(it) }) {
            if (cookingBoost != null) {
                return NutritionProfile(
                    calories = profile.calories * cookingBoost.kcal,
                    protein = profile.protein * cookingBoost.protein,
                    carbs = profile.carbs * cookingBoost.carbs + cookingBoost.additiveCarbs,
                    fats = profile.fats * cookingBoost.fats,
                )
            }
            return profile
        }
    }
    return if (lower.length >= 3) {
        if (cookingBoost != null) {
            NutritionProfile(
                calories = MIXED_DISH.calories * cookingBoost.kcal,
                protein = MIXED_DISH.protein * cookingBoost.protein,
                carbs = MIXED_DISH.carbs * cookingBoost.carbs + cookingBoost.additiveCarbs,
                fats = MIXED_DISH.fats * cookingBoost.fats,
            )
        } else MIXED_DISH
    } else null
}

private data class CookingEstimateBoost(
    val kcal: Double = 1.0,
    val protein: Double = 1.0,
    val carbs: Double = 1.0,
    val fats: Double = 1.0,
    val additiveCarbs: Double = 0.0,
)

private fun detectCookingBoost(foodName: String): CookingEstimateBoost? {
    val lower = foodName.lowercase()
    var boost: CookingEstimateBoost? = null

    if (Regex("""\bfrit[oa]s?\b|\bfritura\b|\bfre[ií]do\b""").containsMatchIn(lower)) {
        boost = CookingEstimateBoost(kcal = 2.0, fats = 3.0)
    }
    if (Regex("""\bempanizad[oa]s?\b|\bapanad[oa]s?\b|\brebozad[oa]s?\b|\btempura\b|\bcapead[oa]s?\b""").containsMatchIn(lower)) {
        boost = CookingEstimateBoost(kcal = 3.0, fats = 3.0, additiveCarbs = 15.0)
    }
    if (Regex("""\bsaltead[oa]s?\b|\bsofrit[oa]s?\b|\bwokead[oa]s?\b""").containsMatchIn(lower)) {
        boost = CookingEstimateBoost(kcal = 1.5, fats = 1.8)
    }
    if (Regex("""\bconfitad[oa]s?\b""").containsMatchIn(lower)) {
        boost = CookingEstimateBoost(kcal = 2.5, fats = 3.5)
    }
    if (Regex("""\bgratinad[oa]s?\b|\bcon\s+queso\s+gratinado\b""").containsMatchIn(lower)) {
        boost = CookingEstimateBoost(kcal = 1.8, fats = 2.0, protein = 1.15)
    }
    if (Regex("""\bplancha\b|\ba\s+la\s+plancha\b""").containsMatchIn(lower)) {
        boost = CookingEstimateBoost(kcal = 1.0, fats = 0.9, protein = 1.05)
    }
    if (Regex("""\bparrilla\b|\basad[oa]s?\s+a\s+la\s+parrilla\b""").containsMatchIn(lower)) {
        boost = CookingEstimateBoost(kcal = 1.05, fats = 0.9, protein = 1.10)
    }

    return boost
}

File: ios-native/KPKNFit/KPKNFit/Domain/Nutrition/NutritionHeuristicEstimator.swift
import Foundation

struct NutritionProfile {
    let calories: Double
    let protein: Double
    let carbs: Double
    let fats: Double
}

private let LEAN_PROTEIN = NutritionProfile(calories: 165, protein: 31, carbs: 0, fats: 3.6)
private let FATTY_PROTEIN = NutritionProfile(calories: 220, protein: 26, carbs: 0, fats: 12)
private let FISH_LEAN = NutritionProfile(calories: 120, protein: 22, carbs: 0, fats: 3)
private let FISH_FATTY = NutritionProfile(calories: 200, protein: 20, carbs: 0, fats: 13)
private let EGG_PROFILE = NutritionProfile(calories: 155, protein: 13, carbs: 1, fats: 11)
private let LEGUME = NutritionProfile(calories: 116, protein: 9, carbs: 20, fats: 0.4)
private let GRAIN_COOKED = NutritionProfile(calories: 130, protein: 3, carbs: 27, fats: 0.5)
private let GRAIN_DRY = NutritionProfile(calories: 360, protein: 13, carbs: 68, fats: 6)
private let PASTA_COOKED = NutritionProfile(calories: 131, protein: 5, carbs: 25, fats: 1.1)
private let BREAD = NutritionProfile(calories: 265, protein: 9, carbs: 49, fats: 3)
private let STARCHY_VEG = NutritionProfile(calories: 85, protein: 1.9, carbs: 20, fats: 0.1)
private let DAIRY_LIQUID = NutritionProfile(calories: 62, protein: 3.2, carbs: 4.8, fats: 3.4)
private let DAIRY_THICK = NutritionProfile(calories: 97, protein: 9, carbs: 3.9, fats: 5)
private let CHEESE = NutritionProfile(calories: 350, protein: 22, carbs: 2, fats: 28)
private let OIL_FAT = NutritionProfile(calories: 780, protein: 0, carbs: 0, fats: 87)
private let NUTS_SEEDS = NutritionProfile(calories: 580, protein: 20, carbs: 20, fats: 50)
private let FRUIT = NutritionProfile(calories: 60, protein: 0.8, carbs: 14, fats: 0.3)
private let VEGETABLE = NutritionProfile(calories: 28, protein: 2, carbs: 5, fats: 0.3)
private let PROTEIN_POWDER = NutritionProfile(calories: 380, protein: 80, carbs: 8, fats: 5)
private let SUGAR_SWEET = NutritionProfile(calories: 380, protein: 0.5, carbs: 90, fats: 0.2)
private let PROCESSED_MEAT = NutritionProfile(calories: 240, protein: 16, carbs: 2, fats: 19)
private let SAUCE_DRESSING = NutritionProfile(calories: 250, protein: 1.5, carbs: 15, fats: 20)
private let MIXED_DISH = NutritionProfile(calories: 160, protein: 10, carbs: 16, fats: 6)

private let KEYWORD_PROFILES: [([String], NutritionProfile)] = [
    (["whey", "proteína en polvo", "proteina en polvo", "caseína", "caseina", "suplemento proteico", "mass gainer", "gainers"], PROTEIN_POWDER),
    (["pechuga de pollo", "pechuga de pavo", "pollo a la plancha", "pechuga", "pechuga cocida", "pollo cocido", "pavo cocido", "clara de huevo", "claras de huevo", "claras", "atún", "atun", "merluza", "tilapia", "corvina", "reineta", "camarón", "camaron", "langostino", "pulpo", "calamar", "pollo", "pavo"], LEAN_PROTEIN),
    (["carne molida", "carne de vacuno", "carne de res", "carne de cerdo", "asado", "costilla", "lomo vetado", "punta de ganso", "tapapecho", "paleta", "pierna de cerdo", "chuleta", "filete", "lomo", "vacuno", "res", "cerdo", "carne"], FATTY_PROTEIN),
    (["salmón", "salmon", "sardina", "caballa", "atún en aceite", "jurel"], FISH_FATTY),
    (["pescado", "congrio", "lenguado", "bacalao"], FISH_LEAN),
    (["huevo duro", "huevo cocido", "huevo frito", "huevo revuelto", "huevos duros", "huevo", "huevos"], EGG_PROFILE),
    (["jamón", "jamon", "salchicha", "vienesa", "longaniza", "mortadela", "salame", "cecina", "tocino", "panceta"], PROCESSED_MEAT),
    (["lentejas", "lenteja", "garbanzos", "garbanzo", "porotos", "poroto", "frijoles", "frijol", "habas", "edamame", "soya"], LEGUME),
    (["pasta", "fideos", "spaghetti", "espagueti", "tallarín", "tallarines", "penne", "macarrón", "macarrones", "lasaña"], PASTA_COOKED),
    (["avena", "granola", "müesli", "muesli", "cereal"], GRAIN_DRY),
    (["arroz blanco", "arroz integral", "arroz", "quinoa", "cebada cocida", "mijo cocido", "trigo bulgur"], GRAIN_COOKED),
    (["marraqueta", "hallulla", "baguette", "pan de molde", "pan integral", "pan blanco", "pan", "tortilla", "arepa", "pita"], BREAD),
    (["papa", "papas", "batata", "camote", "yuca", "ñame", "taro"], STARCHY_VEG),
    (["queso cheddar", "queso gauda", "queso mantecoso", "queso fresco", "queso crema", "queso", "mozzarella", "parmesano"], CHEESE),
    (["yogurt griego", "yogur griego", "yogurt natural", "yogur natural", "yogurt", "yogur", "queso cottage"], DAIRY_THICK),
    (["leche entera", "leche descremada", "leche semi", "leche"], DAIRY_LIQUID),
    (["aceite de oliva", "aceite de coco", "aceite vegetal", "aceite", "mantequilla", "manteca", "ghee", "margarina"], OIL_FAT),
    (["almendra", "almendras", "nuez", "nueces", "maní", "mani", "pecan", "pistache", "semillas de chia", "semillas de linaza", "semillas de girasol", "semillas de zapallo", "semilla", "chía", "chia"], NUTS_SEEDS),
    (["manzana", "plátano", "platano", "banana", "naranja", "uva", "pera", "durazno", "mango", "papaya", "sandía", "sandia", "melón", "melon", "piña", "pina", "frutilla", "frambuesa", "arándano", "arandano", "kiwi", "ciruela", "damasco", "maracuyá", "lúcuma", "fruta"], FRUIT),
    (["lechuga", "espinaca", "brócoli", "brocoli", "zanahoria", "tomate", "cebolla", "pepino", "zapallo", "pimentón", "pimenton", "apio", "rúcula", "repollo", "coliflor", "poroto verde", "choclo", "betarraga", "champiñon", "verdura", "vegetal", "acelga", "alcachofa", "berenjena", "zucchini", "pepinillo"], VEGETABLE),
    (["azúcar", "azucar", "miel", "sirope", "jarabe", "mermelada", "dulce de leche", "manjar", "chocolate"], SUGAR_SWEET),
    (["mayonesa", "ketchup", "mostaza", "salsa de tomate", "salsa de soya", "salsa", "aderezo", "vinagreta"], SAUCE_DRESSING),
]

func estimateNutritionByKeyword(_ foodName: String) -> NutritionProfile? {
    let lower = foodName.trimmingCharacters(in: .whitespaces).lowercased()
    let cookingBoost = detectCookingBoost(lower)

    for (keywords, profile) in KEYWORD_PROFILES {
        if keywords.contains(where: { lower.contains($0) }) {
            if let boost = cookingBoost {
                return NutritionProfile(
                    calories: profile.calories * boost.kcal,
                    protein: profile.protein * boost.protein,
                    carbs: profile.carbs * boost.carbs + boost.additiveCarbs,
                    fats: profile.fats * boost.fats
                )
            }
            return profile
        }
    }

    guard lower.count >= 3 else { return nil }
    if let boost = cookingBoost {
        return NutritionProfile(
            calories: MIXED_DISH.calories * boost.kcal,
            protein: MIXED_DISH.protein * boost.protein,
            carbs: MIXED_DISH.carbs * boost.carbs + boost.additiveCarbs,
            fats: MIXED_DISH.fats * boost.fats
        )
    }
    return MIXED_DISH
}

private struct CookingEstimateBoost {
    let kcal: Double
    let protein: Double
    let carbs: Double
    let fats: Double
    let additiveCarbs: Double
    init(kcal: Double = 1.0, protein: Double = 1.0, carbs: Double = 1.0, fats: Double = 1.0, additiveCarbs: Double = 0.0) {
        self.kcal = kcal; self.protein = protein; self.carbs = carbs; self.fats = fats; self.additiveCarbs = additiveCarbs
    }
}

private let REGEX_FRITO = try! NSRegularExpression(pattern: "\\bfrit[oa]s?\\b|\\bfritura\\b|\\bfre[ií]do\\b")
private let REGEX_EMPANIZADO = try! NSRegularExpression(pattern: "\\bempanizad[oa]s?\\b|\\bapanad[oa]s?\\b|\\brebozad[oa]s?\\b|\\btempura\\b|\\bcapead[oa]s?\\b")
private let REGEX_SALTEADO = try! NSRegularExpression(pattern: "\\bsaltead[oa]s?\\b|\\bsofrit[oa]s?\\b|\\bwokead[oa]s?\\b")
private let REGEX_CONFITADO = try! NSRegularExpression(pattern: "\\bconfitad[oa]s?\\b")
private let REGEX_GRATINADO = try! NSRegularExpression(pattern: "\\bgratinad[oa]s?\\b|\\bcon\\s+queso\\s+gratinado\\b")
private let REGEX_PLANCHA = try! NSRegularExpression(pattern: "\\bplancha\\b|\\ba\\s+la\\s+plancha\\b")
private let REGEX_PARRILLA = try! NSRegularExpression(pattern: "\\bparrilla\\b|\\basad[oa]s?\\s+a\\s+la\\s+parrilla\\b")

private func detectCookingBoost(_ foodName: String) -> CookingEstimateBoost? {
    let range = NSRange(foodName.startIndex..., in: foodName)
    var boost: CookingEstimateBoost? = nil

    if REGEX_FRITO.firstMatch(in: foodName, range: range) != nil {
        boost = CookingEstimateBoost(kcal: 2.0, fats: 3.0)
    }
    if REGEX_EMPANIZADO.firstMatch(in: foodName, range: range) != nil {
        boost = CookingEstimateBoost(kcal: 3.0, fats: 3.0, additiveCarbs: 15.0)
    }
    if REGEX_SALTEADO.firstMatch(in: foodName, range: range) != nil {
        boost = CookingEstimateBoost(kcal: 1.5, fats: 1.8)
    }
    if REGEX_CONFITADO.firstMatch(in: foodName, range: range) != nil {
        boost = CookingEstimateBoost(kcal: 2.5, fats: 3.5)
    }
    if REGEX_GRATINADO.firstMatch(in: foodName, range: range) != nil {
        boost = CookingEstimateBoost(kcal: 1.8, protein: 1.15, fats: 2.0)
    }
    if REGEX_PLANCHA.firstMatch(in: foodName, range: range) != nil {
        boost = CookingEstimateBoost(kcal: 1.0, protein: 1.05, fats: 0.9)
    }
    if REGEX_PARRILLA.firstMatch(in: foodName, range: range) != nil {
        boost = CookingEstimateBoost(kcal: 1.05, protein: 1.10, fats: 0.9)
    }

    return boost
}
